### ISSUE TITLE
Join/leave flood protection can be evaded

### ISSUE THREAD

**Issue by heinrich5991** [teeworlds/teeworlds#1303]
The join/leave flood protection can be evaded using multiple IP addresses.

**Comment by MegaManSec** [comment]
if you block it by username, you can make it so that if you know somebody is entering a server, you can block them.

**Comment by Henningstone** [comment]
@MegaManSec maybe I misunderstood what you meaned, but it is even possible to change the username aswell!

**Comment by Pierre-Sassoulas** [comment]
I think that it is easier to change username than to use multiple ip because that is something you can do in game. Maybe we could use mac adress ?

**Comment by heinrich5991** [comment]
The username is irrelevant to this problem. The MAC address is not transmitted further than your LAN, so this is of no help over the internet (where this problem has relevance).

**Comment by Henningstone** [comment]
The problem is that there is actually no unique and static information sent to the server for exact identification (and recognition) of the clients. All information what the server gets can be changed if you really want to. (Which also might be good for privacy reasons).

**Comment by heinrich5991** [comment]
This is not a problem for basic flood protection.

**Comment by Henningstone** [comment]
That's what I tried to say, and the IP address can be changed, so the circle is closed.
And rather doing it client-sided wouldn't be more than just a kind of workaround, would it?

**Comment by Pierre-Sassoulas** [comment]
Yes I was thinking abot sending the mac to the server from the client, but everything we do on the client can be manipulated by users. A perfect flood protection is impossible imho.

**Comment by heinrich5991** [comment]
Nobody's asking for a "perfect flood protection", we're asking for a better one here.

**Comment by pielover88888** [comment]
Here's an idea. If the server detects lots of rejoin spam, it stops sending it over the chat. It still sends the data to the client about leaving/joining, but does not send it to the chat until it detects the spam stopping or detects an already-joined (stable non rejoinspamming client) leaves. Just for kicks, you could make it attempt to ban the IP address that's spamming for a config value. 'Course, that won't stop the really dedicated-to-spamming people, but it could help with the minor incidents.
Wait, doesn't it already ban for stressing the network, or is that a feature in OpenFNG? Also evadable by VPN/proxy/some other stuff, but it can help. Anything that's better than no protection can be worth something in the long run!

**Comment by Henningstone** [comment]
This idea sounds good to me. I would like to add my ideas to it, though.
Evading your system could be prevented in a way that the server will not send any leave/join chat if it detects e.g. a total of *4 connection changes* per 50 ticks made by less then *4 clients* (it's just an example). It will start sending the chat messages again when at least 2 of the *<4 clients* did not change their connection status for lets say 500 Ticks. This system could ignore clients having a completely new IP address *AND* username, so that innocent players are not affected of it. Anyway, with this the server needs to detect and remember spamming client IPs and usernames, so I know this is improvable too, and I would be grateful if somebody is so kind :)

**Comment by Schwertspize** [comment]
You can use fail2ban to ban everyone with same ip who connect eg 2 times in <1/2 sec so there would be 2 messages and then nothing any more

**Comment by Learath2** [comment]
@Schwertspize fail2ban uses ip addresses. The purpose of this ticket is for multiple ip join flood.

**Comment by oy** [comment]
Doubt there's a sane way to fix this as you can't really distinguish between multiple/one player.
Though sending the information about a connected player to the other clients should be done as late as possible.
And for the record: the clients create the join/leave messages themselves in 0.7.

**Comment by heinrich5991** [comment]
I think @eeeee had a suggestion here.

**Comment by Schwertspize** [comment]
Cool discussion but who have got a botnet under his control and use it ONLY TO FLOOD TEEWORLDS SERVER??? and I don't think you can change vpn/proxy fast enough to really flood...

**Comment by CytraL** [comment]
perhaps adding a idle time when the client sends the petition can mitigate the flood...
Connect Server -> Waiting Time [Pending] -> Join
If X clients are in "waiting time" the server drop the new clients with [many clients connecting simultaneously]...

But... this can cause issues when restart server... xDD

**Comment by BotoX** [comment]
@CytraL Sounds like you can also DoS the server with that (not letting any new players join when it is being flooded).

**Comment by CytraL** [comment]
@BoToX true.... sh*t xDD

**Comment by Henningstone** [comment]
There are clients which can "join" a dummy to the server but instantly disconnect it again and this in an endless loop, which causes servers not to respond or even crash. Talking about (botnet-)DDoS protection is not the actual point imo.

**Comment by Schwertspize** [comment]
I just said: if you are taking about a botnet, then the only way is to only let 1 client per second or something like that connect. if you are talking about one person (with dummy flooding), then I don't understand the multiple ip problem. just quotes:

"The purpose of this ticket is for multiple ip join flood."

"Talking about (botnet-)DDoS protection is not the actual point..."

where is the difference between ddos and multiple ips flooding? (ddos as distributed flooding)

**Comment by BotoX** [comment]
Well, why are we still discussing this, #986 should fix this, right?

**Comment by oy** [comment]
`Well, why are we still discussing this, #986 should fix this, right?`
Nope.
